dotnet/src/Microsoft.Agents.AI.Workflows/Execution/OffThreadRunEventStream.cs

```diff
@@ -0,0 +1,193 @@
+
+    private readonly IInputCoordinator _inputCoordinator;
+    private readonly AsyncCoordinator _outputCoordinator = new();
+
+    private readonly CancellationTokenSource _endRunSource = new();
+    private readonly InitLocked<Task> _runLoopTask = new();
+    private readonly InitLocked<Task> _disposeTask = new();
+
+    private int _isTaken;
+    private int _streamEpoch;
+
+    private readonly ConcurrentQueue<WorkflowEvent> _eventSink = new();
+
+    public OffThreadRunEventStream(ISuperStepRunner stepRunner, IInputCoordinator inputCoordinator)
+    {
+        this.StepRunner = stepRunner;
+
+        this._inputCoordinator = inputCoordinator;
+    }
+
+    private ISuperStepRunner StepRunner { get; }
+
+    public void Start()
+    {
+        this._runLoopTask.Init(() => Task.Run(() => this.RunLoopAsync(), this._endRunSource.Token));
+    }
+
+    public async IAsyncEnumerable<WorkflowEvent> TakeEventStreamAsync([EnumeratorCancellation] CancellationToken cancellation = default)
+    {
+        try
+        {
+            if (Interlocked.Exchange(ref this._isTaken, 1) != 0)
+            {
+                throw new InvalidOperationException("Can only have one active watcher on the event stream at a time.");
+            }
+
+            while (!cancellation.IsCancellationRequested)
+            {
+                if (this._eventSink.TryDequeue(out WorkflowEvent? @event) &&
+                    !cancellation.IsCancellationRequested)
+                {
+                    if (@event is OffThreadHaltSignal haltSignal)
+                    {
+                        int runningEpoch = Volatile.Read(ref this._streamEpoch);
+                        if (haltSignal.Epoch >= runningEpoch)
+                        {
+                            // We hit a halt signal for our current epoch, so we are done.
+                            yield break;
+                        }
+
+                        // We hit a halt signal for a previous epoch, so we ignore it.
+                    }
+                    else
+                    {
+                        yield return @event;
+                    }
+                }
+                else if (!cancellation.IsCancellationRequested)
+                {
+                    await this._outputCoordinator.WaitForCoordinationAsync(cancellation).ConfigureAwait(false);
+                }
+            }
+        }
+        finally
+        {
+            Volatile.Write(ref this._isTaken, 0);
+        }
+    }
+
+    private void NotifyHalt()
+    {
+        int epoch = Math.Max(Volatile.Read(ref this._streamEpoch), 0);
+        this._eventSink.Enqueue(new OffThreadHaltSignal(epoch));
+    }
+
+    private async Task RunLoopAsync(CancellationToken cancellation = default)
+    {
+        this.StepRunner.OutgoingEvents.EventRaised += InspectAndForwardWorkflowEventAsync;
+        bool hadRequestHaltEvent = false;
+
+        int epoch = Volatile.Read(ref this._streamEpoch);
```

Comment:
Is this used anywhere?

Reply:
I think it's meant to be used to break out of the inner/outer loops, but this code is gone for now, to be fixed and restored in a later PR.